proposals/NNNN-module-aliasing-for-disambiguation.md

```diff
@@ -0,0 +1,231 @@
+
+The module alias map will also be used to disallow any references to the binary module names in source files; only the name Utils should appear in source files, not the binary name GameUtils. This is true only if the `-module-alias` was used to build the module (if the renamed modules were directly imported, those names (binary module names) can appear in source files). This restriction is useful as it can make it easier to rename the module again later if needed, e.g. from GameUtils to SwiftGameUtils.   
+
+Unlike source files, the generated interface module (.swiftinterface) will contain the binary module name in all its references. The binary module name will also be stored for indexing and debugging, and treated as the source of truth. 
+
+### Changes to Code Assistance / Indexing
+
+The compiler arguments including the new flag `-module-alias` will be available to SourceKit and indexing. The aliases will be stored in the AST context and used to fetch the right results for code completion and other code assistance features. They will also be stored for indexing so features such as jump to definition can navigate to decls under the binary module names. 
+
+Generated documentation, quick help, and other assistance features will contain the binary module names, which will be treated as the source of truth. 
+
+### Changes to Swift Driver
+
+The module aliasing arguments will be used during dependency scan for both implicit and explicit build modes; the resolved dependency graph will contain the binary module names. In case of the explicit build mode, the dependency input passed to the frontend will contain the binary module names in its json file. Similar to the frontend, validation of the aliasing arguments will be performed at the driver. 
+
+### Changes to SwiftPM
+
+To allow module aliasing more accessible, we will introduce new build configs which can map to the compiler flags for aliasing described above. Let’s go over how they can be adopted by SwiftPM with the second scenario (copied below). 
+```
+App
+|— Module Game (from package ‘swift-game’)
+|— Module Utils (from package ‘swift-game’)
+|— Module Utils (from package ‘swift-draw’)
+```
+Manifest swift-game: the Utils module needs to opt in to allow module aliasing; we will introduce a new parameter called `allowModuleAliasing`. The Utils target has to meet the requirements described in the **Requirements/Limitations** section below. A package can vend multiple targets which might not meet the requirements, so this new parameter needs to be specified per target. If opted in, SwiftPM will perform validations such as whether the target is a pure Swift module.
```

Comment:
This is debatable; I agree that requiring an opt-in can be an extra work and very few will do so. It might be better to just run validations and throw an error if the module doesn't meet the requirements. Will update this section accordingly.